### ISSUE TITLE
Equality among any two top types yields true, including composite types

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#5]
Cf. these SDK issues https://github.com/dart-lang/sdk/issues/33769, https://github.com/dart-lang/sdk/issues/33768, requesting that `Object == dynamic` evaluate to true, and in general that two `Type` entities must be `==` if they both correspond to a top type, and finally that any two types which are subtypes of each other (possibly because they only differ in terms of their choice of top type in certain locations) must also be `==`, e.g., `List<FutureOr<void>>` and `List<FutureOr<FutureOr<dynamic>>>`. But, in all cases, they may not necessarily be `identical`.

We have already said the same thing about `Object` and `void`, in [generalized-void.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/generalized-void.md):

> For a composite type (a generic class instantiation or a function type),
> the reified representation at run time must be such that the type void and
> the built-in class `Object` are treated as equal according to `==`, but they
> need not be `identical`
>.
> [Next, an example is given involving `F<_, _>` showing that it applies also
> with composite types.]

But it seems new to me that we'd do it for `dynamic` as well. So we may need to consider the degree to which this will break existing code.

In favor of doing this, I'd say that it does pretty much amount to treating `dynamic` in the same way that we treat `void`: `dynamic` can be considered to be a variant of `Object` that has an extra (purely static) property that it admits usages (like member lookups or invocation) for which there is no statically known evidence that it will succeed. With `void` I believe that we have agreed many times that it is a variant of `Object` that has an extra (purely static) property that prevents certain usages (in particular, certain data flows), even though the regular semantics would have made the same constructs well-defined and acceptable if `void` had been replaced by `Object` in all the relevant types.

@lrhn has expressed support for this approach. I can see that there is a consistent way to explain why we would do so, and it seems like a nice simplification that we actually only have one top type at run time. @leafpetersen, WDYT?

**Comment by lrhn** [comment]
At run-time there is no distinction between `dynamic`, `void` and `Object`. Both `void` and `dynamic` are static views on `Object`, at run-time they *are* `Object`. We may allow platforms to distinguish the origin of the `Object` type for debugging reasons, but it's the same type. The `Type` objects should reflect this and be equal. 

In general, I'd prefer that `Type.operator==` is defined as: `T1 == T2` iff `T1 <: T2 && T2 <: T1`.

**Comment by eernstg** [comment]
>  `T1 == T2` iff `T1 <: T2 && T2 <: T1`.

This is exactly what the specification has specified since Sep 2018, so the language team has made that decision:

```
When types are reified as instances of the built-in class \code{Type},
those objects override the \lit{==} operator
inherited from the \code{Object} class, so that
two \code{Type} objects are equal according to operator \lit{==}
if{}f the corresponding types are subtypes of each other.
```

So I'll mark this issue as closed by 4816b607a5b74b81a119d2993f770044728eef69.
